### PR TITLE
Speedup node bootstrap after client download

### PR DESCRIPTION
**Pull request by Alenar**
## Content
This PR make the `snapshot download` command of the client append a empty file named 'clean' in the unpacked database.
This is a trick to tell the cardano node to skip the ledger check at startup (this file is usually added by the node itself on a clean shutdown).

## Pre-submit checklist

- Branch
  - [x] Tests are provided (if possible)
  - [x] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested
- Documentation
  - [ ] Update README file (if relevant)
  - [ ] Update documentation website (if relevant)
  - [ ] Add dev blog post (if relevant)

## Issue(s)

Relates to #1131

### CODE REVIEW

**Review by jpraynaud — APPROVED** [review]
LGTM :+1:

File: mithril-client/src/services/snapshot.rs [hunk 1]
```diff
@@ -201,13 +202,13 @@ impl SnapshotService for MithrilClientSnapshotService {
...
         snapshot_entity: &SignedEntity<Snapshot>,
-        pathdir: &Path,
+        download_dir: &Path,
         genesis_verification_key: &str,
         progress_target: ProgressDrawTarget,
     ) -> StdResult<PathBuf> {
         debug!("Snapshot service: download.");
 
-        let unpack_dir = pathdir.join("db");
+        let unpack_dir = download_dir.join("db");
```
Comment (abailly-iohk) [inline mithril-client/src/services/snapshot.rs:211]: suggestion: rename `unpack_dir` to `db_dir`?

File: mithril-client/src/services/snapshot.rs [hunk 2]
```diff
@@ -237,16 +238,30 @@ impl SnapshotService for MithrilClientSnapshotService {
...
 
         progress_bar.println("5/7 - Unpacking the snapshot…")?;
-        let unpacker = unpacker.unpack_snapshot(&filepath, &unpack_dir);
+        let unpacker = unpacker.unpack_snapshot(&snapshot_path, &unpack_dir);
         self.wait_spinner(&progress_bar, unpacker).await?;
 
+        // Append 'clean' file to speedup node bootstrap
+        if let Err(error) = File::create(unpack_dir.join("clean")) {
+            warn!(
+                "Error while creating of a 'clean' file in the unpacked directory to speed\
```
Comment (abailly-iohk) [inline mithril-client/src/services/snapshot.rs:259]: typo: remove `of a`

Comment (abailly-iohk) [reply mithril-client/src/services/snapshot.rs:259]: suggestion: if it's a `warn!` then do not use `Error` in the message or someone looking at the logs might get confused. Rewrite message as something like:
> Could not create clean shutdown marker file in directory {dir}: {err}

File: mithril-client/src/services/snapshot.rs [hunk 3]
```diff
@@ -260,21 +275,25 @@ impl SnapshotService for MithrilClientSnapshotService {
             })?;
 
         progress_bar.println("7/7 - Verifying the snapshot signature…")?;
-        let mut protocol_message = certificate.protocol_message.clone();
-        protocol_message.set_message_part(
-            ProtocolMessagePartKey::SnapshotDigest,
-            unpacked_snapshot_digest,
-        );
-        if protocol_message.compute_hash() != certificate.signed_message {
+        let expected_message = {
```
Comment (abailly-iohk) [inline mithril-client/src/services/snapshot.rs:278]: This change seems unrelated to the PR's purpose ?

File: mithril-client/src/services/snapshot.rs [hunk 4]
```diff
@@ -260,21 +275,25 @@ impl SnapshotService for MithrilClientSnapshotService {
...
+            );
+            protocol_message.compute_hash()
+        };
+
+        if expected_message != certificate.signed_message {
             debug!("Digest verification failed, removing unpacked files & directory.");
 
-            if let Err(e) = std::fs::remove_dir_all(&unpack_dir) {
-                warn!("Error while removing unpacked files & directory: {e}.");
+            if let Err(error) = std::fs::remove_dir_all(&unpack_dir) {
```
Comment (abailly-iohk) [inline mithril-client/src/services/snapshot.rs:290]: Same here, but I get the point of renaming a one-letter variable :)

File: mithril-client/src/services/snapshot.rs [hunk 5]
```diff
@@ -472,40 +531,65 @@ mod tests {
...
+            .expect("Snapshot download should succeed.");
+        assert!(
+            filepath.is_dir(),
+            "Unpacked location must be in a directory."
+        );
+        assert_eq!(Some(OsStr::new("db")), filepath.file_name());
+    }
+
+    #[tokio::test]
+    async fn test_download_snapshot_ok_add_clean_file_allowing_node_bootstrap_speedup() {
```
Comment (abailly-iohk) [inline mithril-client/src/services/snapshot.rs:570]: Do we have a test checking _not adding_ the 'clean' file still works? I was under the impression the could would always try to create it.

**Review by abailly-iohk — COMMENTED** [review]
There appears to be some unrelated changes in formatting which clutter the PR, but otherwise LGTM.

File: mithril-client/src/services/snapshot.rs (hunk 3 above)
Comment (Alenar) [reply mithril-client/src/services/snapshot.rs:278]: Yes, I did a "boy scoot" on the file clean up before working on this issue. The bulk of the code implementing the issue is on the second commit "Make client append a empty 'clean' file after download succeed".

File: mithril-client/src/services/snapshot.rs (hunk 5 above)
Comment (Alenar) [reply mithril-client/src/services/snapshot.rs:564]: I would be nice bust costly , to do that cleanly I need to extract the code that insert the `clean` file in a separated service to mock it. For such a small part I don't think that's worth the increased complexity.
